hazelcast-jet-core/src/main/java/com/hazelcast/jet/function/Observer.java

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.function;
+
+import com.hazelcast.jet.Observable;
+
+import javax.annotation.Nonnull;
+import java.util.function.Consumer;
+
+/**
+ * Watcher of the events produced by an {@link Observable}. Once subscribed, it
+ * will be notified of all past events produced by the {@link Observable}
+ * (to the extent to which these haven't yet been overwritten in the
+ * {@link Observable} implementation's finite history) and
+ * all future events that will be produced by the {@link Observable}, while
+ * the {@link Observer} remains added to it.
+ * <p>
+ * Notification will be done by an internal pool of threads and care should
+ * be taken to not block those threads, ie. finish processing of all events
+ * as soon as possible.
+ *
+ * @param <T> type of data values in the sequence produced by the
+ * {@link Observable}
+ */
+@FunctionalInterface
+public interface Observer<T> {
+
+    /**
+     * Utility method for building an {@link Observer} from its basic
+     * components.
+     */
+    static <T> Observer<T> of(
+            @Nonnull Consumer<? super T> onNext,
+            @Nonnull Consumer<? super Throwable> onError,
+            @Nonnull Runnable onComplete
+    ) {
+        return new Observer<T>() {
+            @Override
+            public void onNext(T t) {
+                onNext.accept(t);
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                onError.accept(throwable);
+            }
+
+            @Override
+            public void onComplete() {
+                onComplete.run();
+            }
+        };
+    }
+
+    /**
+     * Method that will be called when data values from the {@link Observable}
+     * become available.
+     * <p>
+     * The data values passed via this method don't have a clear,
+     * global ordering. Some {@link Observable}s for example produce values
+     * in parallel so the order in which they arrive at the
+     * {@link Observer}s is unpredictable.
+     * <p>
+     * It is not possible to observe data values after an error or completion
+     * has been observed.
+     */
+    void onNext(@Nonnull T t);
+
+    /**
+     * Method that will be called when the {@link Observable} (to which this
+     * {@link Observer} is subscribed to) encounters an error in its
+     * process of producing data values.
+     * <p>
+     * The passed {@link Throwable} instance attempts to reflect/explain the
+     * original error as closely as possible.
+     * <p>
+     * Once an error has been observed, no further data values nor completion
+     * events will be received.
+     */
+    default void onError(@Nonnull Throwable throwable) {
+        throwable.printStackTrace();
```

Comment:
is this a good default behaviour? or should we just rethrow it?

Reply:
if we rethrow, it will be ignored. Or it should be...

Reply:
Yes, if I rethrow it then that will be an exception flying out our async thread pool (or client executor service). Isn't printing it better? 

I personally would prefer to not have a default `onError` implementation at all and always force the user to write one.